akka-typed/src/main/scala/akka/typed/cluster/ClusterSingleton.scala

```diff
@@ -0,0 +1,218 @@
+  val dataCenter:                      Option[DataCenter],
+  val singletonIdentificationInterval: FiniteDuration,
+  val removalMargin:                   FiniteDuration,
+  val handOverRetryInterval:           FiniteDuration,
+  val bufferSize:                      Int) extends NoSerializationVerificationNeeded {
+
+  /**
+   * Replace the required roles with a new set of required roles
+   */
+  def withRole(role: String): ClusterSingletonSettings =
+    new ClusterSingletonSettings(Some(role), dataCenter, singletonIdentificationInterval, removalMargin, handOverRetryInterval, bufferSize)
+
+  def withNoRole(): ClusterSingletonSettings =
+    new ClusterSingletonSettings(None, dataCenter, singletonIdentificationInterval, removalMargin, handOverRetryInterval, bufferSize)
+
+  def withDataCenter(dataCenter: DataCenter): ClusterSingletonSettings =
+    new ClusterSingletonSettings(None, Some(dataCenter), singletonIdentificationInterval, removalMargin, handOverRetryInterval, bufferSize)
+
+  def withNoDataCenter(): ClusterSingletonSettings =
+    new ClusterSingletonSettings(None, None, singletonIdentificationInterval, removalMargin, handOverRetryInterval, bufferSize)
+
+  def withRemovalMargin(removalMargin: FiniteDuration): ClusterSingletonSettings =
+    new ClusterSingletonSettings(role, dataCenter, singletonIdentificationInterval, removalMargin, handOverRetryInterval, bufferSize)
+
+  def withHandoverRetryInterval(handOverRetryInterval: FiniteDuration): ClusterSingletonSettings =
+    new ClusterSingletonSettings(role, dataCenter, singletonIdentificationInterval, removalMargin, handOverRetryInterval, bufferSize)
+
+  def withBufferSize(bufferSize: Int): ClusterSingletonSettings =
+    new ClusterSingletonSettings(role, dataCenter, singletonIdentificationInterval, removalMargin, handOverRetryInterval, bufferSize)
+
+  /**
+   * INTERNAL API:
+   */
+  @InternalApi
+  private[akka] def toManagerSettings(singletonName: String): ClusterSingletonManagerSettings =
+    new ClusterSingletonManagerSettings(singletonName, role, removalMargin, handOverRetryInterval)
+
+  /**
+   * INTERNAL API:
+   */
+  @InternalApi
+  private[akka] def toProxySettings(singletonName: String): ClusterSingletonProxySettings =
+    new ClusterSingletonProxySettings(singletonName, role, singletonIdentificationInterval, bufferSize)
+
+  /**
+   * INTERNAL API:
+   */
+  @InternalApi
+  private[akka] def shouldRunManager(cluster: Cluster): Boolean =
+    (role.isEmpty || cluster.selfMember.roles(role.get)) &&
+      (dataCenter.isEmpty || dataCenter.contains(cluster.selfMember.dataCenter))
+
+}
+
+object ClusterSingleton extends ExtensionId[ClusterSingleton] {
+
+  override def createExtension(system: ActorSystem[_]): ClusterSingleton = new ClusterSingletonImpl(system)
+
+}
+
+/**
+ * INTERNAL API:
+ */
+@InternalApi
+private[akka] object ClusterSingletonImpl {
+  def managerNameFor(singletonName: String) = s"singleton-manager-${singletonName}"
```

Comment:
so we're sticking to `-` in names? Making sure all names are consistent, so far I think we've rather had `singletonManager` names hm

Reply:
I also vote for `singletonManager` and I think that is what I have used for most cluster things, but I know there are other conventions in other places :(

Reply:
I'll change it to camel case.